File: Hot 100/543. 二叉树的直径.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int ans=0;
    int depth(TreeNode* root){
        if(root == nullptr){
            return 0;
        }
        int L=depth(root->left);
        int R=depth(root->right);
        ans = max(ans, L+R+1);
        return max(L, R)+1;
    }
    int diameterOfBinaryTree(TreeNode* root) {
        ans=1;
        depth(root);
        return ans-1;
    }
};

File: 刷题/Hot 100/543. 二叉树的直径.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 func diameterOfBinaryTree(root *TreeNode) int {
    ans := 1
    var depth func(*TreeNode) int
    depth = func (root *TreeNode) int {
        if root == nil {
            return 0
        }
        L := depth(root.Left)
        R := depth(root.Right)
        ans = max(ans, L + R + 1)
        return max(L, R)+1
    }
    depth(root)
    return ans-1
}
func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}